llama_index/agent/openai_agent.py

```diff
@@ -53,6 +55,38 @@ def call_function(
     )
 
 
+class StreamingChatResponse:
+    """Streaming chat response to user and writing to chat history."""
+
+    def __init__(self, chat_stream: ChatResponseGen) -> None:
+        self.chat_stream = chat_stream
+        self.queue: Queue = Queue()
+        self.is_done = False
+        self.response = ""
+
+    def write_response_to_history(self, chat_history: List[ChatMessage]) -> None:
+        final_message = None
+        for chat in self.chat_stream:
+            final_message = chat.message
+            self.queue.put_nowait(chat.delta)
+
+        if final_message is not None:
+            chat_history.append(final_message)
+
+        self.is_done = True
+
+    @property
+    def response_gen(self) -> Generator[str, None, None]:
+        while not self.is_done or not self.queue.empty():
+            try:
+                delta = self.queue.get(block=False)
+                self.response += delta
+                yield delta
+            except Exception:
+                # Queue is empty, but we're not done yet
```

Comment:
Isn't this a possibility if we read off the queue faster than the client writes into it?

Reply:
How does this logic compare to `llama_index/langchain_helpers/streaming.py`?

Reply:
it's possible, which is why the loop only ends when both `is_done` is set and the queue is empty